### DISCUSSION TITLE
i can not research related words for a Chinese word in CNW

### DISCUSSION THREAD

**Discussion by xy1137030414** [goodmami/wn#198] [Q&A] [answered]
is there a way that i can research related words （like，antonym，similar）for a Chinese word？

**Comment by fcbond** [comment, marked as answer]
Please take a look at https://github.com/goodmami/wn/discussions/161#discussioncomment-2448312 where I show how to find antonyms for Japanese.  

It should be easy to adapt to other languages and other relationships.

**Comment by xy1137030414** [comment]
thans you so much，but i still meet some problems，when i process my data，here are my coding and track：

`# 遍历词列表
for i, (word, pos) in enumerate(words):
    # 获取同义词集
    synsets = zh.synsets(word, pos=pos)

    if not synsets:
        print("Error: synsets list is empty or not defined.")
        continue
    print(f"Found {len(synsets)} synsets for word '{word}'")  # 打印找到的同义词集数量

    for synset in synsets:
    # 获取ILI
        ili = synset.ili
        print(type(ili))
        #print(dir(ili))
        if not isinstance(ili, str):
            print(f"ILI id of synset {synset} is {ili.id}.")
            print(f"ILI definition of synset {synset} is {ili.definition}.")
            continue
        if ili is not None and isinstance(ili, str):
            # 在oewn:2023中获取英文同义词集
            eng_synsets = wn.synsets(ili=ili, lexicon='oewn:2023')
            for eng_synset in eng_synsets:
                # 获取定义、上位词、下位词、meronym、holoym，antonym、synonym和similar的同义词集
                definition = eng_synset.definition()
                hypernyms = eng_synset.hypernyms()
                hyponyms = eng_synset.hyponyms()
                meronyms = eng_synset.part_meronyms()
                holonyms = eng_synset.part_holonyms()
                antonyms = eng_synset.lemmas()[0].antonyms() if eng_synset.lemmas() else []
                synonyms = eng_synset.lemma_names()
                similars = eng_synset.similar_tos()
                # 添加到结果列表
                results.append([i+1, word, synset, ili, definition, synonyms, hypernyms, hyponyms, meronyms, holonyms, antonyms, similars])

print(f"Found {len(results)} results")  # 打印找到的结果数量`

`track：ead` 5 words from 'word.txt'
Found 2 synsets for word '牛'
ILI id of synset Synset('omw-cmn-02402425-n') is i48233.
ILI definition of synset Synset('omw-cmn-02402425-n') is <bound method ILI.definition of ILI('i48233')>.
ILI id of synset Synset('omw-cmn-02403003-n') is i48234.
ILI definition of synset Synset('omw-cmn-02403003-n') is <bound method ILI.definition of ILI('i48234')>.
Found 1 synsets for word '高兴'
ILI id of synset Synset('omw-cmn-07527352-n') is i76310.
ILI definition of synset Synset('omw-cmn-07527352-n') is <bound method ILI.definition of ILI('i76310')>.
Found 2 synsets for word '工作'
ILI id of synset Synset('omw-cmn-02594979-v') is i34660.
ILI definition of synset Synset('omw-cmn-02594979-v') is <bound method ILI.definition of ILI('i34660')>.
ILI id of synset Synset('omw-cmn-02410855-v') is i33794.
ILI definition of synset Synset('omw-cmn-02410855-v') is <bound method ILI.definition of ILI('i33794')>.
Found 8 synsets for word '工作'
ILI id of synset Synset('omw-cmn-00409483-n') is i37553.
ILI definition of synset Synset('omw-cmn-00409483-n') is <bound method ILI.definition of ILI('i37553')>.
ILI id of synset Synset('omw-cmn-00795720-n') is i39587.
ILI definition of synset Synset('omw-cmn-00795720-n') is <bound method ILI.definition of ILI('i39587')>.
ILI id of synset Synset('omw-cmn-00620752-n') is i38687.
ILI definition of synset Synset('omw-cmn-00620752-n') is <bound method ILI.definition of ILI('i38687')>.
ILI id of synset Synset('omw-cmn-00584367-n') is i38467.
ILI definition of synset Synset('omw-cmn-00584367-n') is <bound method ILI.definition of ILI('i38467')>.
ILI id of synset Synset('omw-cmn-00719705-n') is i39217.
ILI definition of synset Synset('omw-cmn-00719705-n') is <bound method ILI.definition of ILI('i39217')>.
ILI id of synset Synset('omw-cmn-00582388-n') is i38458.
ILI definition of synset Synset('omw-cmn-00582388-n') is <bound method ILI.definition of ILI('i38458')>.
ILI id of synset Synset('omw-cmn-03599628-n') is i55272.
ILI definition of synset Synset('omw-cmn-03599628-n') is <bound method ILI.definition of ILI('i55272')>.
ILI id of synset Synset('omw-cmn-13968092-n') is i110340.
ILI definition of synset Synset('omw-cmn-13968092-n') is <bound method ILI.definition of ILI('i110340')>.
Found 3 synsets for word '幸福'
ILI id of synset Synset('omw-cmn-14447525-n') is i112931.
ILI definition of synset Synset('omw-cmn-14447525-n') is <bound method ILI.definition of ILI('i112931')>.
ILI id of synset Synset('omw-cmn-13987719-n') is i110441.
ILI definition of synset Synset('omw-cmn-13987719-n') is <bound method ILI.definition of ILI('i110441')>.
ILI id of synset Synset('omw-cmn-13987423-n') is i110440.
ILI definition of synset Synset('omw-cmn-13987423-n') is <bound method ILI.definition of ILI('i110440')>.
Found 0 results
Results saved to `'results.txt'`

could you find the problems about my code？thank you very much！

**Comment by fcbond** [comment]
Hi,

I'm not sure what you are trying to do.   

But one error is that `definition` is a method, so should be `definition()`

**Comment by xy1137030414** [comment]
Hi，Fcbond！i am so sorry that i was not introduce what i want to do.
My goal is researching lexical chian in Chinese texts.
To achieve this,  I should use wn this python lib to get synset for each line Chinese word in txt（this line include a lemma and pos）.
Second, I want to   konw the synset of every chinese word  in zh.
third, for geting the sematic relationships of chinese word , i must use ili to get the same english synset,and find its hypernym,meronym,holonym,etc.
Final，all results should save to a new .txt.

In my code,  i think the errors can be find in this part:
for synset in synsets:
    # 获取ILI
        ili = synset.ili
        print(type(ili))
        #print(dir(ili))
        if not isinstance(ili, str):
            print(f"ILI id of synset {synset} is {ili.id}.")
            print(f"Error: ili of synset {synset} is not a string.")
            print(f"ILI definition of synset {synset} is {ili.definition}.")
            continue
        if ili is not None and isinstance(ili, str):
            # 在oewn:2023中获取英文同义词集
            eng_synsets = wn.synsets(ili=ili, lexicon='oewn:2023')
            for eng_synset in eng_synsets:
                # 获取定义、上位词、下位词、meronym、holoym，antonym、synonym和similar的同义词集
                definition = eng_synset.definition()
                hypernyms = eng_synset.hypernyms()
                hyponyms = eng_synset.hyponyms()
                meronyms = eng_synset.part_meronyms()
                holonyms = eng_synset.part_holonyms()
                antonyms = eng_synset.lemmas()[0].antonyms() if eng_synset.lemmas() else []
                synonyms = eng_synset.lemma_names()
                similars = eng_synset.similar_tos()
                # 添加到结果列表
                results.append([i+1, word, synset, ili, definition, synonyms, hypernyms, hyponyms, meronyms, holonyms, antonyms, similars])

print(f"Found {len(results)} results")  # 打印找到的结果数量

# 保存到文本文件
with open('results.txt', 'w', encoding='utf-8') as f:
    for result in results:
        f.write('\t'.join(str(r) for r in result) + '\n')

print("Results saved to 'results.txt'")  # 打印保存结果的消息
 and i also print the one of content in terminal:
Read 5 words from 'word.txt'
[Synset('omw-cmn-02402425-n'), Synset('omw-cmn-02403003-n')]
Found 2 synsets for word '牛'
<class 'wn.ILI'>
ILI id of synset Synset('omw-cmn-02402425-n') is i48233.
Error: ili of synset Synset('omw-cmn-02402425-n') is not a string.
ILI definition of synset Synset('omw-cmn-02402425-n') is <bound method ILI.definition of ILI('i48233')>.

So the reason is ili is not a string?
I am so sorry to bother you,!

________________________________

**Comment by goodmami** [comment]
Hi @xy1137030414, it looks like you are having trouble looking up related synsets for some `omw-cmn` synsets.

First, the `Synset.ili` property is a [ILI](https://wn.readthedocs.io/en/latest/api/wn.html#the-ili-class) object, not a string, so when you do:

```python
        ili = synset.ili
        ...
        if ili is not None and isinstance(ili, str):
            ...
```

That `isinstance(ili, str)` condition will never be true and the code block will not execute. You can get the string from `ili.id`, as you have done elsewhere. See the docs for [Using Interlingual Indices](https://wn.readthedocs.io/en/latest/guides/interlingual.html#using-interlingual-indices) for more info.

But you should not have to do this manual step as Wn will handle most of the technical details. For example:

```pycon
>>> cmn = wn.Wordnet('omw-cmn', expand='omw-en')
>>> cmn.synsets("牛")
[Synset('omw-cmn-02402425-n'), Synset('omw-cmn-02403003-n')]
>>> cmn.synsets("牛")[0].hypernyms()  # Wn will look up hypernyms in omw-en via the ILI
[Synset('omw-cmn-02402010-n')]
>>> cmn.synsets("牛")[0].hypernyms()[0].lemmas()
['牛属动物']
```

You can read more about this from the [Cross-lingual Relation Traversal](https://wn.readthedocs.io/en/latest/guides/interlingual.html#cross-lingual-relation-traversal) docs.

Finally, I have some other recommendations:
* Ensure that you have an English lexicon (e.g., `omw-en` or `oewn`) installed. The relations for OMW lexicons come from an English lexicon.
* Create a `Wordnet` object as I have done above, using an explicit `expand='...'` argument (where `...` is the English lexicon you have installed). If unspecified, it may not be clear which, if any, expand lexicons are used for relation traversal.
